cloud_storage/cloud_storage/overrides/file.py

```diff
@@ -141,30 +141,69 @@ def associate_files(
 			existing_file.attached_to_doctype = attached_to_doctype
 			existing_file.attached_to_name = attached_to_name
 			existing_file.append(
-				"file_association", add_child_file_association(attached_to_doctype, attached_to_name)
+				"file_association",
+				add_child_file_association(attached_to_doctype, attached_to_name),
 			)
 			existing_file.save()
 		else:
 			if self.file_association:
 				link_names = [i.link_name for i in self.file_association]
 				if attached_to_name not in link_names:
 					self.append(
-						"file_association", add_child_file_association(attached_to_doctype, attached_to_name)
+						"file_association",
+						add_child_file_association(attached_to_doctype, attached_to_name),
 					)
 			else:
 				self.append(
-					"file_association", add_child_file_association(attached_to_doctype, attached_to_name)
+					"file_association",
+					add_child_file_association(attached_to_doctype, attached_to_name),
 				)
 
 	def add_file_version(self, version_id):
-		self.append(
-			"versions",
-			{
-				"version": str(version_id),
-				"user": frappe.session.user,
-				"timestamp": get_datetime(),
-			},
+		previous_files = frappe.get_all(
+			"File",
+			filters={"file_name": self.file_name, "name": ["!=", self.name], "is_folder": 0},
+			fields=["name"],
+			order_by="creation desc",
 		)
+		if previous_files:
+			self.append(
+				"custom_versions",
+				{
+					"version": str(version_id),
+					"user": frappe.session.user,
+					"timestamp": get_datetime(),
+				},
+			)
+
+			for each in previous_files:
+				file_doc = frappe.get_doc("File", each.name)
+				for assoc in file_doc.file_association:
+					self.append(
+						"file_association",
+						add_child_file_association(assoc.link_doctype, assoc.link_name),
+					)
+
+				for version in file_doc.custom_versions:
+					self.append(
+						"custom_versions",
```

Comment:
In the Custom Field doctype you can rename the fieldname. 
`frappe.rename_doc('Custom Field', 'File-custom_versions', 'File-version')`
This should be included in the cloud_storage/custom/file.json